### PR TITLE
Feature chroma support per block LR specification

### PR DESCRIPTION
**Pull request by maybleMyers**
This enables you to specify per block learning rate multipliers. Chroma suffers from lines and other abnormalities if you train certain blocks too much. This is an example addition to your launch command with known good values:

--network_args 'lr_if_contains=double_blocks$$0$$:0' 'lr_if_contains=double_blocks$$1$$:0' 'lr_if_contains=double_blocks$$2$$:0.00456' 'lr_if_contains=double_blocks$$3$$:0.006' 'lr_if_contains=double_blocks$$4$$:0.00722' 'lr_if_contains=double_blocks$$5$$:0.00822' 'lr_if_contains=double_blocks$$6$$:0.009' 'lr_if_contains=double_blocks$$7$$:0.00956' 'lr_if_contains=double_blocks$$8$$:0.00989' 'lr_if_contains=double_blocks$$9$$:0.01' 'lr_if_contains=double_blocks$$10$$:0.00989' 'lr_if_contains=double_blocks$$11$$:0.00956' 'lr_if_contains=double_blocks$$12$$:0.009' 'lr_if_contains=double_blocks$$13$$:0.00822' 'lr_if_contains=double_blocks$$14$$:0.00722' 'lr_if_contains=double_blocks$$15$$:0.006' 'lr_if_contains=double_blocks$$16$$:0.00456' 'lr_if_contains=double_blocks$$17$$:0' 'lr_if_contains=double_blocks$$18$$:0'

### CODE REVIEW

**Comment by kohya-ss** [comment]
Thank you, I think this is a very useful feature!

However, SDXL LoRA in sd-scripts already has a function to specify the learning rate per block, which is an option called `block_lr`. I think it would be better to unify the notation.

Also, in existing LoRA, the way to specify rank and alpha per block differed depending on the architecture (SD, SDXL, FLUX). I would like to be able to specify this using a unified notation (for example, a regular expression-based notation) in future, so I'd like to adopt the future-proof notation at this time.

I will consider how to implement this, so I would appreciate it if you could wait for a while.

**Comment by kohya-ss** [comment]
I would like to implement the learning rate and dim (rank) specification using regular expressions using the following function. What do you think?

---

### 6.4. Regular Expression-based Rank/LR Configuration / 正規表現によるランク・学習率の指定

You can specify ranks (dims) and learning rates for LoRA modules using regular expressions. This allows for more flexible and fine-grained control than specifying by layer.

These settings are specified via the `network_args` argument.

*   `network_reg_dims`: Specify ranks for modules matching a regular expression. The format is a comma-separated string of `pattern=rank`.
    *   Example: `--network_args "network_reg_dims=single.*_mod.*=4,img_attn=8"`
    *   This sets the rank to 4 for modules whose names start with `single` and contain `_mod`, and to 8 for modules containing `img_attn`.
*   `network_reg_lrs`: Specify learning rates for modules matching a regular expression. The format is a comma-separated string of `pattern=lr`.
    *   Example: `--network_args "network_reg_lrs=single_blocks.*=1e-5,double_blocks.*=5e-5"`
    *   This sets the learning rate to `1e-5` for modules whose names start with `single_blocks`, and to `5e-5` for modules whose names start with `double_blocks`.

**Notes:**

*   Settings via `network_reg_dims` and `network_reg_lrs` take precedence over the global `--network_dim` and `--learning_rate` settings.
*   If a module name matches multiple patterns, the setting from the last matching pattern in the string will be applied.
*   These settings are applied after the block-specific training settings (`train_double_block_indices`, `train_single_block_indices`).

**Comment by maybleMyers** [comment]
I think your solution is better xD. It will be a more condensed command and can accomplish all of the same things. It will allow more variations as well. It would be nice if we could also use it for fine tuning and not just lora.

**Comment by kohya-ss** [comment]
> I think your solution is better xD. It will be a more condensed command and can accomplish all of the same things. It will allow more variations as well. It would be nice if we could also use it for fine tuning and not just lora.

Thank you for verification!

I've added the feature. Please see [the doc](https://github.com/kohya-ss/sd-scripts/blob/feature-chroma-support/docs/flux_train_network.md#65-regular-expression-based-ranklr-configuration--%E6%AD%A3%E8%A6%8F%E8%A1%A8%E7%8F%BE%E3%81%AB%E3%82%88%E3%82%8B%E3%83%A9%E3%83%B3%E3%82%AF%E5%AD%A6%E7%BF%92%E7%8E%87%E3%81%AE%E6%8C%87%E5%AE%9A) for details.

I'd like to close this PR.
